README.md
# Ch4Demo
 

api/CHANGELOG.md
### [1.0.3](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v1.0.2...v1.0.3) (2023-02-05)


### Bug Fixes

* **deps:** update dependency au.com.dius.pact.provider:junit5 to v4.4.5 ([#131](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/131)) ([d9ee209](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/d9ee209dcb9e148130d6eb4ae0ab489394dd18a4))

### [1.0.2](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v1.0.1...v1.0.2) (2023-02-04)


### Bug Fixes

* **deps:** update dependency org.jetbrains.kotlin:kotlin-stdlib to v1.8.10 ([#130](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/130)) ([d2f709f](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/d2f709fa2862f4c370ee0be946dd77c5d728cf4d))

### [1.0.1](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v1.0.0...v1.0.1) (2023-02-01)


### Bug Fixes

* **deps:** update dependency com.fasterxml.jackson.core:jackson-databind to v2.14.2 ([#127](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/127)) ([c63c0b7](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/c63c0b73f691087874bb7d2c15196db26fe3ec2e))

## [1.0.0](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.10...v1.0.0) (2023-01-21)


### ⚠ BREAKING CHANGES

* replace exampleCompany with liatrio

### Features

* replace exampleCompany with liatrio ([7eabfd3](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/7eabfd369eaf29cab13c2dd53eee356bfb2c5c45))


### Bug Fixes

* **deps:** update dependency au.com.dius.pact.provider:junit5 to v4.4.4 ([#124](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/124)) ([2ba927d](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/2ba927d0db88f91d69cffa27a0b6c46a928850c0))
* **deps:** update spring boot to v2.7.8 ([#125](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/125)) ([a117e46](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/a117e4620e0c1c38f6ffdb60f4ddd68d534b9cc9))

### [0.4.10](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.9...v0.4.10) (2023-01-19)


### Bug Fixes

* add default gateway.enabled value ([99bfe78](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/99bfe78dc1a83dc473f2ca58082ffddcedfbf38f))
* **deps:** update dependency org.apache.tomcat.embed:tomcat-embed-core to v9.0.71 ([#121](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/121)) ([f60251b](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/f60251bb80ce94e904c90290f08b7af04858a7c1))
* **deps:** update spring core to v5.3.25 ([#119](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/119)) ([7cf1cc6](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/7cf1cc64fd0d72e32bfac172ee7b5c5a33d3d890))

### [0.4.9](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.8...v0.4.9) (2023-01-14)


### Bug Fixes

* **deps:** update junit5 monorepo ([#117](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/117)) ([6b637c7](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/6b637c7d2a48320d0e88678bce10924eb096be7a))
* **deps:** update spring data ([#120](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/120)) ([63d7bee](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/63d7bee6c6ceb8e491eb3bddcba524e1ec6da35f))

### [0.4.8](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.7...v0.4.8) (2023-01-08)


### Bug Fixes

* **deps:** update dependency com.google.code.gson:gson to v2.10.1 ([#115](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/115)) ([47d0611](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/47d0611d8d2937afeb5293869f807387f0a326a8))

### [0.4.7](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.6...v0.4.7) (2022-12-31)


### Bug Fixes

* **deps:** update dependency org.mockito:mockito-core to v4.11.0 ([#112](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/112)) ([b73790e](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/b73790e1c00e18fd1fd9b9b01b015c74e8290547))

### [0.4.6](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.5...v0.4.6) (2022-12-29)


### Bug Fixes

* **deps:** update dependency org.jetbrains.kotlin:kotlin-stdlib to v1.8.0 ([#111](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/111)) ([99c1b1b](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/99c1b1b8f628fb1df2802eebc6efa3c9f3323f47))

### [0.4.5](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.4...v0.4.5) (2022-12-24)


### Bug Fixes

* **deps:** update dependency au.com.dius.pact.provider:junit5 to v4.4.3 ([#110](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/110)) ([62b1392](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/62b1392c70805b2311d52aab8946043fca480192))

### [0.4.4](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.3...v0.4.4) (2022-12-24)


### Bug Fixes

* **deps:** update spring boot to v2.7.7 ([#109](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/issues/109)) ([7754f26](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/commit/7754f2605f748c42f9e203be9b9db7c537b4b3db))

### [0.4.3](https://github.com/liatrio-delivery-povs/devops-knowledge-share-api/compare/v0.4.2...v0.4.3) (2022-12-19)


### Bug Fixes


api/src/test/java/com/liatrio/dojo/devopsknowledgeshareapi/PostControllerTest.java
package com.liatrio.dojo.devopsknowledgeshareapi;

import static org.mockito.MockitoAnnotations.initMocks;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup;

import org.json.JSONObject;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

@SpringBootTest
@AutoConfigureMockMvc()
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class PostControllerTest {
    @Autowired
    private MockMvc mockMvc;

    @InjectMocks
    PostController mockPost;

    @Mock
    PostRepository mockPr;

    private JSONObject testPost;

    @BeforeAll
    public void setup() throws Exception {
        initMocks(this);
        this.mockMvc = standaloneSetup(mockPost).build();

        testPost = new JSONObject();

        testPost.put("firstName", "John");
        testPost.put("title", "My First Post");
        testPost.put("link", "https://www.example.com/blog/post-1");
    }

    @Test
    public void getPostsResponse() throws Exception {
        this.mockMvc.perform(get("/posts")).andDo(print()).andExpect(status().isOk());
    }

    @Test
    public void postPostsFirstName() throws Exception {
        this.mockMvc.perform(
                post("/posts").content(testPost.toString()).with(csrf()).contentType(MediaType.APPLICATION_JSON))
                .andDo(print()).andExpect(status().isOk());
    }

    @Test
    public void deletePostsFirstName() throws Exception {
        this.mockMvc.perform(
                post("/posts").content(testPost.toString()).with(csrf()).contentType(MediaType.APPLICATION_JSON))
                .andDo(print()).andExpect(status().isOk());
        this.mockMvc.perform(delete("/posts/1/").with(csrf())).andDo(print()).andExpect(status().isOk());
    }
}

api/src/functionalTest/java/com/liatrio/dojo/devopsknowledgeshareapi/pojo/BlogPost.java
package com.liatrio.dojo.devopsknowledgeshareapi.pojo;

public class BlogPost {
    private int objectId;
    private String firstName;
    private String title;
    private String link;
    private String datePosted;
    private String imageUrl;
    private String dateAsDate;
    private int id;

    public int getObjectId() {
        return objectId;
    }

    public void setObjectId(int objectId) {
        this.objectId = objectId;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getLink() {
        return link;
    }

    public void setLink(String link) {
        this.link = link;
    }

    public String getDatePosted() {
        return datePosted;
    }

    public void setDatePosted(String datePosted) {
        this.datePosted = datePosted;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getDateAsDate() {
        return dateAsDate;
    }

    public void setDateAsDate(String dateAsDate) {
        this.dateAsDate = dateAsDate;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }


}

api/src/functionalTest/java/com/liatrio/dojo/devopsknowledgeshareapi/functional/PropertiesExtractor.java
package com.liatrio.dojo.devopsknowledgeshareapi.functional;

import java.io.FileInputStream;
import java.io.IOException;
import java.net.URL;
import java.util.Properties;

public class PropertiesExtractor {
    private static Properties properties;
    private static String properties_file = System.getenv("spring_profiles_active") != null ? System.getenv("spring_profiles_active") : "application";
    static {
        properties = new Properties();
        URL url = new PropertiesExtractor().getClass().getClassLoader().getResource(properties_file+".properties");
        try{
            properties.load(new FileInputStream(url.getPath()));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static String getProperty(String key){
        return properties.getProperty(key);
    }
}


api/src/performanceTest/java/com/liatrio/dojo/devopsknowledgeshareapi/simulations/ExampleSimulation.java
package com.liatrio.dojo.devopsknowledgeshareapi.simulations;

import static io.gatling.javaapi.core.CoreDsl.RawFileBody;
import static io.gatling.javaapi.core.CoreDsl.constantUsersPerSec;
import static io.gatling.javaapi.core.CoreDsl.global;
import static io.gatling.javaapi.core.CoreDsl.holdFor;
import static io.gatling.javaapi.core.CoreDsl.jumpToRps;
import static io.gatling.javaapi.core.CoreDsl.scenario;
import static io.gatling.javaapi.http.HttpDsl.http;
import static io.gatling.javaapi.http.HttpDsl.status;

import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;
import io.gatling.javaapi.http.HttpProtocolBuilder;
import java.time.Duration;

public class ExampleSimulation extends Simulation {
  String baseUrl = TestConfiguration.BASE_URL;

  HttpProtocolBuilder httpProtocol = http
    .warmUp(baseUrl + "/actuator/health")
    .baseUrl(baseUrl);

  ScenarioBuilder baseScenario = scenario("Validation for health check")
    .exec(http("/actuator/health")
      .get("/actuator/health")
      .check(status().in(200)));
  
  {
    setUp(
      baseScenario.injectOpen(constantUsersPerSec(TestConfiguration.PARALLEL_QUERIES).during(Duration.ofMinutes(TestConfiguration.RAMP_DURATION))))
      .throttle(
        jumpToRps(TestConfiguration.RAMP_TO),
        holdFor(Duration.ofMinutes(TestConfiguration.RAMP_DURATION))
      )
      .protocols(httpProtocol)
      .assertions(
        global().successfulRequests().percent().gt(TestConfiguration.SUCCESS_PERCENTAGE)
      );
  }
}


api/src/functionalTest/java/com/liatrio/dojo/devopsknowledgeshareapi/functional/HealthCheckTest.java
package com.liatrio.dojo.devopsknowledgeshareapi.functional;


import com.liatrio.dojo.devopsknowledgeshareapi.pojo.BlogPost;
import io.restassured.builder.RequestSpecBuilder;
import io.restassured.filter.log.ErrorLoggingFilter;
import io.restassured.filter.log.RequestLoggingFilter;
import io.restassured.filter.log.ResponseLoggingFilter;
import io.restassured.path.json.JsonPath;
import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.jupiter.api.*;
import org.springframework.boot.test.context.SpringBootTest;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintStream;

import static io.restassured.RestAssured.given;

@SpringBootTest
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class HealthCheckTest {

    private static PropertiesExtractor propertiesExtractor;
    public static PrintStream printStream;
    public static RequestSpecification request;
    public static int blogId;
    static {
        try {

            //For extracting constant endpoints and constant url's from properties file
            propertiesExtractor = new PropertiesExtractor();
            //For the purpose of Logging
            printStream = new PrintStream(new FileOutputStream("logging.txt"), true);
            //Preparing the Common Spec request for POST,GET, DELETE
            request = new RequestSpecBuilder()
                    //.setBaseUri("http://localhost:8080")
                    .setBaseUri(propertiesExtractor.getProperty("spring.application.url"))
                    .setBasePath("/posts")
                    .addHeader("Content-Type", "application/json; charset=UTF-8")
                    .addHeader("Accept", "application/json, text/javascript, */*; q=0.01")
                    .addHeader("Connection", "keep-alive")
                    .addFilter(RequestLoggingFilter.logRequestTo(printStream))
                    .addFilter(ResponseLoggingFilter.logResponseTo(printStream))
                    .addFilter(ErrorLoggingFilter.logErrorsTo(printStream))
                    .build();

        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    /**
     * Purpose: API request to perform an Independent 'GET' operation without depending on other methods
     * Outcome: Creates one BlogPost and Get's the list of Blogposts and deletes the blogPost at the end
     */
    @Test
    @Order(2)
    public void getRequest() throws IOException, JSONException {
        //Create a new blog post to run this method independent of another
        Response blogResponse = createBlogPost();
        JsonPath jsonPath = new JsonPath(blogResponse.asString());
        blogId = jsonPath.get("id");
        //Performing the Get Request for created Blog Id
        Response response = given().spec(request).get();
        JSONArray JSONResponseBody = new JSONArray(response.body().asString());
        for (int i = 0; i < JSONResponseBody.length(); i++) {
            JSONObject jsonobject = JSONResponseBody.getJSONObject(i);
            if (jsonobject.getInt("id") == blogId) {
                String name = jsonobject.getString("firstName");
                String title = jsonobject.getString("title");
                String link = jsonobject.getString("link");
                String imageUrl = jsonobject.getString("imageUrl");
                int id = jsonobject.getInt("id");
                Assertions.assertEquals("liatrio", name);
                Assertions.assertEquals("Testing", title);
                Assertions.assertEquals("http://www.liatrio.com", link);
                Assertions.assertEquals("http://www.liatrio.image.com", imageUrl);
                Assertions.assertEquals(blogId, id);
                //To check the output in command line terminal with gradle build you can uncomment and use it
                /* System.out.println("****************************");
                    System.out.println("Name::" + name);
                    System.out.println("title::" + title);
                    System.out.println("link::" + link);
                    System.out.println("****************************");
                */
            }//End If
        }//End For
        //Deleting the created Blog Post for Re-Usability and clear the database with un-necessary blog id's
        deleteBlogPost(blogId);
    }// End Method

    /**
     * Purpose: API request to perform a 'POST' operation for creating a BlogPost independently without depending on other
     * methods
     * Outcome: Creates a new BlogPost and grabs a created blogId and removes the record at the end for database cleanup
     */
    @Test
    @Order(1)
    public void  postRequest()  {
        //Leaving the below code commented to help the beginners how the request body is passed
        // as Json object in 'blogpost' pojo class by using Jackson,gson,johnzon libraries
        /*  String requestBody = "{\n" +
                "  \"firstName\": \"test2\",\n" +
                "  \"title\": \"string2\",\n" +
                "  \"link\": \"http://liatrio.com\",\n" +
                "  \"datePosted\": \"2022-04-07T16:28:30.813Z\",\n" +
                "  \"imageUrl\": \"http://liatrio.com\",\n" +
                "  \"dateAsDate\": \"2022-04-07T16:28:30.813Z\"\n" +
                "}";*/
        Response blogResponse = createBlogPost();
        Assertions.assertEquals(200, blogResponse.getStatusCode());
        JsonPath jsonPath = new JsonPath( blogResponse.asString());
        blogId = jsonPath.get("id");
        deleteBlogPost(blogId);
    }

    /**
     * Purpose: API request to perform a 'DELETE' operation in Deleting a BlogPost independently with out depending
     * on other methods
     * Outcome: Delete's the selected BlogId
     */
    @Test
    @Order(3)
    public void deleteRequest()  {
        //creating and Independent BlogPost
        Response blogResponse = createBlogPost();
        JsonPath jsonPath = new JsonPath( blogResponse.asString());
        blogId = jsonPath.get("id");
        //Grabbing the BlogId and deleting it
        Response  response = given().spec(request).delete(String.valueOf(blogId));
        Assertions.assertEquals(200, response.statusCode());
    }

    /**
     * Purpose: Re-Usable function for the purpose of creating Blogpost
     * @return Respsone Object
     */
    public Response createBlogPost(){
        BlogPost blogPost = new BlogPost();
        blogPost.setFirstName("liatrio");
        blogPost.setTitle("Testing");
        blogPost.setLink("http://www.liatrio.com");
        blogPost.setDatePosted("2022-04-12T19:47:43.915Z");
        blogPost.setImageUrl("http://www.liatrio.image.com");
        blogPost.setDateAsDate("2022-04-12T19:47:43.915Z");
        Response  response = given().spec(request).body(blogPost).post();
        return response;
    }

    /**
     * Purpose: Re-Usable function for the purpose of deleting blogId
     * @param blogId
     * @return blogId
     */
    public int deleteBlogPost(int blogId){
        Response  response = given().spec(request).delete(String.valueOf(blogId));
        return blogId;
    }


}

